### PR TITLE
CIV-8639 - Any Document Upload API endpoint

### PR DESCRIPTION
**Pull request by m-meulendijks-v1**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/CIV-8639

### Change description ###
After uploading a document as MultiPartFile, the following case document is returned from Civil-Service (no document type):
    {
        "documentLink": {
            "document_url": "http://dm-store:8080/documents/9f97df48-bbae-4bb6-8f12-256d44796a83",
            "document_binary_url": "http://dm-store:8080/documents/9f97df48-bbae-4bb6-8f12-256d44796a83/binary",
            "document_filename": "Check Scheduler.png",
            "document_hash": "846bfa880e75a45dc66ff5ba9517584e934a25d7154adb7d8481ff3e52002ee5"
        },
        "documentName": "Check Scheduler.png",
        "documentSize": 14772,
        "createdDatetime": "2023-06-01T11:05:50",
        "createdBy": "Civil"
    }

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/SecuredDocumentManagementService.java [hunk 1]
```diff
@@ -98,6 +99,58 @@ public CaseDocument uploadDocument(String authorisation, PDF pdf) {
...
+                                                                 .toInstant()
+                                                                 .atZone(ZoneId.systemDefault())
+                                                                 .toLocalDateTime()))
+                .documentSize(document.size)
+                .createdBy(CREATED_BY)
+                .build();
+
+        } catch (Exception ex) {
+            log.error("Failed uploading file {}", originalFileName, ex);
+            throw new DocumentUploadException(originalFileName, ex);
```
Comment (LeonardoPalmeiro) [inline commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/SecuredDocumentManagementService.java:149]: the DocumentUploadException should be add into ControllerAdvice if not then we will get always 500 from controller side, some times from DM store we can get http 404, 403 and so on.

File: src/main/java/uk/gov/hmcts/reform/civil/controllers/cases/DocumentController.java [hunk 2]
```diff
@@ -27,13 +33,27 @@
...
         @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation, @NotNull @RequestBody CaseData caseData) {
         return claimFormService.uploadSealedDocument(authorisation, caseData);
     }
 
+    @PostMapping(value = "/generateAnyDoc")
+    @Operation(summary = "Upload document")
+    public CaseDocument uploadAnyDocument(
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
+        @RequestParam("file")MultipartFile file) {
+        try {
```
Comment (LeonardoPalmeiro) [inline src/main/java/uk/gov/hmcts/reform/civil/controllers/cases/DocumentController.java:49]: my suggestion in here is:
public ResponseEntity<CaseDocument>uploadAnyDocument(
        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
        @RequestParam("file")MultipartFile file) {
UploadedDocument uploadedDocument = new UploadedDocument(file.getOriginalFilename(), file.getBytes());
            return ResponseEntity.ok(documentManagementService.uploadDocument(authorisation, uploadedDocument);
            }

File: commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/UnsecuredDocumentManagementService.java [hunk 3]
```diff
@@ -91,6 +93,55 @@ public CaseDocument uploadDocument(String authorisation, PDF pdf) {
...
+            UploadResponse response = documentUploadClientApi.upload(
+                authorisation,
+                authTokenGenerator.generate(),
+                userInfo.getUid(),
+                documentManagementConfiguration.getUserRoles(),
+                Classification.RESTRICTED,
+                singletonList(file)
+            );
+
+            Document document = response.getEmbedded().getDocuments().stream()
```
Comment (LeonardoPalmeiro) [inline commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/UnsecuredDocumentManagementService.java:119]: may we don't need this part.
If we get some problem in this documentUploadClientApi.upload method will throw some error. I believe so.
please could you check what happened when we have error on that point

**Review by LeonardoPalmeiro — CHANGES_REQUESTED** [review]

**Review by kurt-uk — CHANGES_REQUESTED** [review]

**Comment by kurt-uk** [comment]
nice work! just need a tweak on the object returned from the controller and its a good opportunity to introduce http error handling to the upload controller that for some reason was missed out before

File: src/main/java/uk/gov/hmcts/reform/civil/controllers/cases/DocumentController.java [hunk 4]
```diff
@@ -27,13 +33,27 @@
...
+            UploadedDocument uploadedDocument = new UploadedDocument(file.getOriginalFilename(), file.getBytes());
+            return documentManagementService.uploadDocument(authorisation, uploadedDocument);
```
[the remark this answers is not in the document]
Comment (m-meulendijks-v1) [reply src/main/java/uk/gov/hmcts/reform/civil/controllers/cases/DocumentController.java]: I think I've addressed your issue - implemented it much like you suggested.

File: src/main/java/uk/gov/hmcts/reform/civil/controllers/cases/DocumentController.java [hunk 5]
```diff
@@ -27,13 +33,27 @@
...
 
     @PostMapping("/generateSealedDoc")
     public CaseDocument uploadSealedDocument(
         @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation, @NotNull @RequestBody CaseData caseData) {
         return claimFormService.uploadSealedDocument(authorisation, caseData);
     }
 
+    @PostMapping(value = "/generateAnyDoc")
+    @Operation(summary = "Upload document")
+    public CaseDocument uploadAnyDocument(
```
[the remark this answers is not in the document]
Comment (m-meulendijks-v1) [reply src/main/java/uk/gov/hmcts/reform/civil/controllers/cases/DocumentController.java]: I think I've addressed your issue - implemented it much like you suggested.

File: commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/UnsecuredDocumentManagementService.java (hunk 3 above)
Comment (m-meulendijks-v1) [reply commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/UnsecuredDocumentManagementService.java:118]: The error generally occurs at documentUploadClientApi.upload() - I've not seen it fail at this later point.
When it fails at documentUploadClientApi.upload(), I generally get this exception message back:
_org.camunda.bpm.extension.rest.exception.RemoteProcessEngineException: REST-CLIENT-001 Error during remote Camunda engine invocation of CaseDocumentClientApi#uploadDocuments(String,String,DocumentUploadRequest):_

**Review by LeonardoPalmeiro — APPROVED** [review]

File: commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/SecuredDocumentManagementService.java [hunk 6]
```diff
@@ -98,6 +99,57 @@ public CaseDocument uploadDocument(String authorisation, PDF pdf) {
...
+                .createdDatetime(LocalDateTimeHelper.fromUTC(document.createdOn
+                                                                 .toInstant()
+                                                                 .atZone(ZoneId.systemDefault())
+                                                                 .toLocalDateTime()))
+                .documentSize(document.size)
+                .createdBy(CREATED_BY)
+                .build();
+
+        } catch (Exception ex) {
+            log.error("Failed uploading file {}", originalFileName, ex);
```
Comment (kurt-uk) [inline commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/SecuredDocumentManagementService.java:147]: now the controller advice has the error log we dont need here too, file name could be passed in to the DocumentUploadException and pulled when needed

File: src/main/java/uk/gov/hmcts/reform/civil/controllers/cases/DocumentController.java [hunk 7]
```diff
@@ -27,11 +34,26 @@
...
+    }
+
+    @PostMapping(value = "/generateAnyDoc")
+    @Operation(summary = "Upload document")
+    public ResponseEntity<CaseDocument> uploadAnyDocument(
+        @RequestHeader(HttpHeaders.AUTHORIZATION) String authorisation,
+        @RequestParam("file")MultipartFile file) {
+
+        UploadedDocument uploadedDocument = new UploadedDocument(file.getOriginalFilename(), file);
+        return ResponseEntity
```
Comment (kurt-uk) [inline src/main/java/uk/gov/hmcts/reform/civil/controllers/cases/DocumentController.java:54]: 👍🏼

**Review by kurt-uk — CHANGES_REQUESTED** [review]

**Review by LeonardoPalmeiro — APPROVED** [review]

File: commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/UnsecuredDocumentManagementService.java
[the remark this answers is not in the document]
Comment (m-meulendijks-v1) [reply commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/UnsecuredDocumentManagementService.java:1]: Put extra status code checks in the tests.

File: commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/SecuredDocumentManagementService.java (hunk 6 above)
Comment (m-meulendijks-v1) [reply commons-civil/src/main/java/uk/gov/hmcts/reform/civil/documentmanagement/SecuredDocumentManagementService.java]: Removed the duplicate log

**Review by kurt-uk — APPROVED** [review]

**Pull request merged**